lib/FrontendTool/FrontendTool.cpp

```diff
@@ -314,47 +311,43 @@ static bool writeSIL(SILModule &SM, const PrimarySpecificPaths &PSPs,
                   PSPs.OutputFilename, opts.EmitSortedSIL);
 }
 
-static bool printAsObjCIfNeeded(StringRef outputPath, ModuleDecl *M,
-                                StringRef bridgingHeader, bool moduleIsPublic) {
-  using namespace llvm::sys;
-
+/// A wrapper around swift::atomicallyWritingToFile that handles diagnosing any
+/// filesystem errors and ignores empty output paths.
+static bool atomicallyWritingToTextFile(
+    StringRef outputPath, DiagnosticEngine &diags,
+    llvm::function_ref<bool(llvm::raw_pwrite_stream &)> action) {
   if (outputPath.empty())
     return false;
 
-  clang::CompilerInstance Clang;
-
-  std::string tmpFilePath;
-  std::error_code EC;
-  std::unique_ptr<llvm::raw_pwrite_stream> out =
-    Clang.createOutputFile(outputPath, EC,
-                           /*Binary=*/false,
-                           /*RemoveFileOnSignal=*/true,
-                           /*BaseInput=*/"",
-                           path::extension(outputPath),
-                           /*UseTemporary=*/true,
-                           /*CreateMissingDirectories=*/false,
-                           /*ResultPathName=*/nullptr,
-                           &tmpFilePath);
-
-  if (!out) {
-    M->getASTContext().Diags.diagnose(SourceLoc(), diag::error_opening_output,
-                                      tmpFilePath, EC.message());
-    return true;
-  }
-
-  auto requiredAccess = moduleIsPublic ? AccessLevel::Public
-                                       : AccessLevel::Internal;
-  bool hadError = printAsObjC(*out, M, bridgingHeader, requiredAccess);
-  out->flush();
-
-  EC = swift::moveFileIfDifferent(tmpFilePath, outputPath);
+  bool actionFailed = false;
+  std::error_code EC =
+      swift::atomicallyWritingToFile(outputPath, /*binary*/false,
+                                     [&](llvm::raw_pwrite_stream &out) {
+    actionFailed = action(out);
+  });
```

Comment:
I think it might read a little more clearly with an early return before the `if (EC)`:
```
if (!actionFailed)
  return false;
```
Then, you could move the `return true;` down to the last line of the function.

Reply:
Sigh... You are right. If the action fails (`action` returns true--BTW `action` return convention (true == error) ought to also be in the Doxygen, sigh) the file is still written. Is this right? Should the file be removed if `action` returns true?

Reply:
I think yes, it's a temporary file and it would clutter up the user's directory.

(None of our actions currently fail, so this is kind of a moot point. Another option would be to take that part out.)

Reply:
If we don't remove the file, at least we could include an assertion that the action succeeds with a FIXME about removing the file. Or something...

Reply:
Oh, wait, hang on. Are you talking about removing the temporary file or the final output file? As written the code always cleans up the temporary file, but does not delete the final output file if there's an error in writing. I figured that was a safer option given Clang's mentioning "you can write to this file but not to the directory that contains it", although it does sometimes mean you've produced erroneous data.

Reply:
If line 326 above evaluates to true, whatever file is written by atomicallyWritingToFile is left on the file system, whichever one that is. That’s what I saw.

Reply:
Right, okay. That's intended but I could see how we might prefer otherwise.